pkg/volume/csi/nodeinfomanager/nodeinfomanager.go

```diff
@@ -0,0 +1,450 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package nodeinfomanager includes internal functions used to add/delete labels to
+// kubernetes nodes for corresponding CSI drivers
+package nodeinfomanager // import "k8s.io/kubernetes/pkg/volume/csi/nodeinfomanager"
+
+import (
+	"encoding/json"
+	"fmt"
+
+	csipb "github.com/container-storage-interface/spec/lib/go/csi/v0"
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/util/retry"
+	csiv1alpha1 "k8s.io/csi-api/pkg/apis/csi/v1alpha1"
+	csiclientset "k8s.io/csi-api/pkg/client/clientset/versioned"
+)
+
+const (
+	// Name of node annotation that contains JSON map of driver names to node
+	annotationKeyNodeID = "csi.volume.kubernetes.io/nodeid"
+)
+
+var nodeKind = v1.SchemeGroupVersion.WithKind("Node")
+
+// nodeInfoManager contains necessary common dependencies to update node info on both
+// the Node and CSINodeInfo objects.
+type nodeInfoManager struct {
+	nodeName      types.NodeName
+	k8s           kubernetes.Interface
+	csiKubeClient csiclientset.Interface
+}
+
+// If no updates is needed, the function must return the same Node object as the input.
+type nodeUpdateFunc func(*v1.Node) (newNode *v1.Node, updated bool, err error)
+
+// Interface implements an interface for managing labels of a node
+type Interface interface {
+	// Record in the cluster the given node information from the CSI driver with the given name.
+	AddNodeInfo(driverName string, driverNodeID string, topology *csipb.Topology) error
+
+	// Remove in the cluster node information from the CSI driver with the given name.
+	RemoveNodeInfo(driverName string) error
+}
+
+// NewNodeInfoManager initializes nodeInfoManager
+func NewNodeInfoManager(
+	nodeName types.NodeName,
+	kubeClient kubernetes.Interface,
+	csiKubeClient csiclientset.Interface) Interface {
+	return &nodeInfoManager{
+		nodeName:      nodeName,
+		k8s:           kubeClient,
+		csiKubeClient: csiKubeClient,
+	}
+}
+
+// AddNodeInfo updates the node ID annotation in the Node object and CSIDrivers field of the
+// CSINodeInfo object.
+func (nim *nodeInfoManager) AddNodeInfo(driverName string, driverNodeID string, topology *csipb.Topology) error {
```

Comment:
Should it be locked somehow? Is it possible that two drivers gets registered at the same time and AddNodeInfo is called twice?

Reply:
Yep it's entirely possible 2 drivers are registered at the same time - socket fsnotify events are handled in parallel.

None of the node info manger's methods mutate any internal state. However multiple calls will cause retries due to Node/CSINodeInfo conflict. Even though kubelet ends up making more calls than needed, it might be safer than using locks, which could potentially block progress due to network calls.

I'll call this out in a comment

Reply:
However in the future when plugin unregistration is implemented, there could be a race between `AddNodeInfo()` and `RemoveNodeInfo()` on creating and deleting `CSINodeInfo`...

Reply:
We could cross that bridge when we get there, since plugin unregistration implementation might already have mechanisms in place to serialize register and unregister calls.